controllers/tools.js

```diff
@@ -231,31 +237,253 @@ const showTools = (req, res) => {
       baseUrl: '/tools/?p={{page}}' + sortQuery + limitQuery
     };
 
-    res.render('tools/tools', {title: 'Tools', head, body, pagination, user: res.locals.currentUser, limit: true,
-      themeTitle: process.env.SC_NAV_TITLE || 'Schul-Cloud', versions, messageTypes, privacies, authMethods});
+    res.render('tools/tools', {
+        title: 'Tools',
+        head,
+        body,
+        pagination,
+        user: res.locals.currentUser,
+        limit: true,
+        themeTitle: process.env.SC_NAV_TITLE || 'Schul-Cloud',
+        versions,
+        messageTypes,
+        privacies,
+        authMethods
+    });
   });
 };
 
-// secure routes
-router.use(authHelper.authChecker);
+// CTL tool handlers
+
+const mapToolToOauthClient = (body, create = false) => {
+    const { config } = body;
+    return {
+        client_id: create ? config.clientId : undefined,
+        client_name: body.name,
+        client_secret: config.clientSecret,
+        redirect_uris: config.redirectUris.split(";"), // missing
+        token_endpoint_auth_method: config.tokenEndpointAuthMethod, // missing
+        subject_type: "pairwise", // maybe missing
+        scope: config.scope, // missing
+        frontchannel_logout_uri: config.frontchannelLogoutUri,
+    };
+};
+
+const mapOauthClientToOauthConfig = (client) => {
+    return {
+        clientId: client.client_id,
+        clientSecret: '',
+        redirectUris: client.redirect_uris.join(";"),
+        tokenEndpointAuthMethod: client.token_endpoint_auth_method,
+        scope: client.scope,
+        frontchannelLogoutUri: client.frontchannel_logout_uri,
+    };
+};
+
+const sanitizeCtlTool = (req, create=false) => {
+    req.body.key = req.body.key || null;
+    req.body.resource_link_id = req.body.resource_link_id || null;
+    req.body.lti_version = req.body.lti_version || null;
+    req.body.lti_message_type = req.body.lti_message_type || null;
+    req.body.secret = (req.body.secret === PASSWORD ? undefined : req.body.secret);
+    req.body.isTemplate = true;
+    if(create || !req.body.isLocal) { // non-local (LTI) tools can be updated forever, local (OAuth2) only during creation
+        req.body.oAuthClientId = req.body.oAuthClientId || "";
+    } else {
+        req.body.oAuthClientId = undefined; // undefine the property prohibits database update
+    }
+    req.body.isLocal = (create ? !!req.body.isLocal : undefined);
```

Comment:
Can we rename isLocal in new Tools or maybe delete it? When it is true we call hydra directly after tool creation or update.. i think this should be done in the server